config.txt
menu_show_mod_menu: 0
target_monitor: 
gamepad_3_look_sensitivity: 1.25
type_enabled_visible__terrain_type: 1
type_enabled_visible__spawn_point: 1
type_enabled_visible__rigged_object: 1
type_enabled_visible__reflection_capture_object: 1
type_enabled_visible__prefab: 1
type_enabled_visible__placeholder_object: 1
type_enabled_visible__path_point_object: 1
type_enabled_visible__navmesh_region_object: 0
type_enabled_visible__navmesh_hint_object: 0
type_enabled_visible__navmesh_connection_object: 0
gamepad_3_deadzone: 0.075
type_enabled_visible__movement_object: 1
type_enabled_visible__light_volume_object: 1
gamepad_2_deadzone: 0.075
type_enabled_visible__light_probe_object: 1
difficulty_preset_value: 3
type_enabled_visible__item_object: 1
type_enabled_visible__hotspot_object: 1
type_enabled_visible__env_object: 1
screenwidth: 1280
gamepad_1_look_sensitivity: 1.25
type_enabled_visible__editable_entity_group: 1
gamepad_0_look_sensitivity: 1.25
gamepad_1_deadzone: 0.075
playername: Glimpse
type_enabled_visible__dynamic_light_object: 1
type_enabled_visible__decal_object: 1
type_enabled_visible__ambient_sound_object: 1
type_enabled_edit__terrain_type: 1
type_enabled_edit__movement_object: 0
type_enabled_edit__spawn_point: 1
type_enabled_edit__rigged_object: 1
type_enabled_edit__reflection_capture_object: 1
type_enabled_edit__prefab: 0
gamepad_0_deadzone: 0.075
type_enabled_edit__placeholder_object: 0
menu_show_log: 0
type_enabled_edit__path_point_object: 0
level_history1: D:/Apps/Steam/SteamFileData/steamapps/common/Overgrowth/Data/Levels/og_story/01_Slaver_Camp_Landing.xml
menu_show_warnings: 0
menu_show_state: 0
menu_show_sound: 0
menu_show_save: 0
menu_show_performance: 0
menu_show_mp_settings: 0
menu_show_mp_info: 0
menu_show_mp_debug: 0
type_enabled_edit__navmesh_region_object: 0
gamepad_2_look_sensitivity: 1.25
menu_show_input_debug: 0
type_enabled_edit__navmesh_hint_object: 0
type_enabled_edit__navmesh_connection_object: 0
type_enabled_edit__light_volume_object: 1
type_enabled_edit__light_probe_object: 0
type_enabled_edit__item_object: 0
type_enabled_edit__hotspot_object: 0
level_history9: 
type_enabled_edit__env_object: 0
type_enabled_edit__editable_entity_group: 0
type_enabled_edit__dynamic_light_object: 1
type_enabled_edit__decal_object: 0
level_history8: 
level_history7: 
level_history6: 
level_history5: 
level_history4: 
level_history3: D:/Apps/Steam/SteamFileData/steamapps/common/Overgrowth/Data/Levels/t2/s1/b5.xml
level_history2: D:/Apps/Steam/SteamFileData/steamapps/common/Overgrowth/Data/Levels/nothing.xml
level_history10: 
type_enabled_edit__camera_type: 1
type_enabled_edit__ambient_sound_object: 1
custom_resolution: 0
type_enabled_visible__camera_type: 1
screenheight: 720
fullscreen: false
imgui_scale: 0
vsync: 0
limit_fps_in_game: false
max_frame_rate: 60
simple_shadows: 1
blood: 2
blood_color: 0.4 0 0
multisample: 1
anisotropy: 4
motion_blur_amount: 0
post_effects: true
shaders: true
texture_reduce: 0
texture_minimize_ram: 1
full_level_unload: true
gamma_correct: true
fps_label: false
sound_label: false
visible_raycasts: false
visible_sound_spheres: false
editor_mode: false
invert_x_mouse_look: false
invert_y_mouse_look: false
use_raw_input: true
tet_mesh_lighting: 0
light_volume_lighting: false
global_time_scale_mult: 1
game_difficulty: 1
tutorials: 0
difficulty_set: false
volume_shadows: false
ssao: false
auto_ledge_grab: 0
particle_field: true
particle_field_simple: false
custom_level_shaders: true
disable_fog: false
simple_fog: false
depth_of_field: true
depth_of_field_reduced: false
attrib_envobj_instancing: false
no_auto_nav_mesh: false
no_texture_convert: false
gl_load_s3tc: true
skip_loading_pause: false
brightness: 1
use_soundtrack_limiter: true
master_volume: 0.8
music_volume: 0.4
baked_shadows: true
debug_key_presses: false
block_cheating_progress: true
auto_camera: false
split_screen: false
chase_camera_fov: 90.0
editor_camera_fov: 90.0
report_case_errors: false
background_process_pool: false
main_menu: true
debug_load_level: Project60/16_red_desert.xml
mouse_sensitivity: 0.5
gpu_skinning: false
detail_objects: 1
detail_objects_reduced: 0
seamless_cubemaps: true
debug_keys: true
shader_dir_path: Data/GLSL/
debug_show_ai_state: false
debug_show_ai_path: false
debug_draw_window: false
albedo_only: false
no_decals: false
single_pass_shadow_cascade: false
shader_debug: false
opengl_callback_errors: true
opengl_callback_error_dialog: false
occlusion_query: true
gamma_correct_final_output: true
dump_include_scripts: false
allow_game_dir_save: false
level_backup_count: 10
no_reflection_capture: 0
has_detected_settings: 1
save_as_crunch: false
check_for_workshop_membership: true
enable_live_update: true
level_load_stress: false
detail_object_decals: 1
asdebugger_enabled: 0
asdebugger_break_on_script_change: 0
menu_show_asdebugger_contexts: 0
asprofiler_enabled: 0
menu_show_asprofiler: 0
language: en_us
key[quit]: esc
key[slow]: tab
key[fast]: \
key[crouch]: lshift
key[jump]: space
key[show_timing]: f2
key[screenshot]: f7
key[transparent_screenshot]: f8
key[console]: `
key[left]: a
key[right]: d
key[up]: w
key[down]: s
key[item]: e
key[drop]: q
key[skip_dialogue]: return
key[rclick]: rshift
key[attack]: mouse0
key[grab]: mouse2
key[walk]: lctrl
key[menu_up]: up
key[menu_down]: down
key[menu_left]: left
key[menu_right]: right
key[menu_return]: return
key[debug_revive_all_characters]: x
key[debug_cut_throat]: ,
key[debug_lightning]: f
key[debug_make_injured_ragdoll]: n
key[debug_make_limp_ragdoll]: m
key[debug_make_ragdoll]: z
key[debug_scream]: v
key[debug_path]: p
key[debug_misc_action]: b
key[debug_switch_to_combat_rabbit]: 1
key[debug_switch_to_civ_rabbit]: 2
key[debug_switch_to_cat]: 3
key[debug_switch_to_rat]: 4
key[debug_switch_to_wolf]: 5
key[debug_switch_to_dog]: 6
key[debug_switch_to_rabbot]: 7
gamepad[up]: lefty-
gamepad[down]: lefty+
gamepad[left]: leftx-
gamepad[right]: leftx+
gamepad[look_up]: righty-
gamepad[look_down]: righty+
gamepad[look_left]: rightx-
gamepad[look_right]: rightx+
gamepad[jump]: rightshoulder
gamepad[crouch]: leftshoulder
gamepad[attack]: righttrigger
gamepad[grab]: lefttrigger
gamepad[item]: a
gamepad[drop]: x
gamepad[quit]: back
gamepad[skip_dialogue]: start
gamepad[menu_up]: dpup
gamepad[menu_down]: dpdown
gamepad[menu_left]: dpleft
gamepad[menu_right]: dpright
gamepad[menu_return]: a
menu_player_config: 0
gamepad_deadzone: 0.075
gamepad_look_sensitivity: 1.25
controller[movex]: axis1
controller[movey]: axis2
controller[lookx]: axis3
controller[looky]: axis4
controller[jump]: button11
controller[crouch]: button10
controller[item]: button1
controller[drop]: button3
controller[attack]: axis6>0.5
controller[grab]: axis5>0.5
bind[pause]: `
bind_win[quit]: alt+f4
bind_unix[quit]: cmd+q
bind[back]: esc
bind[clone_transform]: alt
bind[snap_transform]: cmd
bind[force_rotate]: r
bind[force_scale]: e
bind[force_translate]: t
bind[normal_transform]: shift
bind[edit_script_params]: u
bind[single_selected]: i
bind[edit_color]: cmd+p
bind[search_scenegraph]: cmd+f
bind[scenegraph]: y
bind[toggle_player]: cmd+shift+p
bind[toggle_object_editing]: cmd+1
bind[toggle_decal_editing]: cmd+2
bind[toggle_hotspot_editing]: cmd+3
bind[toggle_light_editing]: cmd+4
bind[open_spawner]: cmd+l
bind[save_level]: cmd+s
bind[save_level_as]: cmd+shift+s
bind[save_selected_items]: cmd+alt+s
bind[cut]: cmd+x
bind[paste]: cmd+v
bind[copy]: cmd+c
bind[undo]: cmd+z
bind[redo]: cmd+shift+z
bind[enable_imposter]: cmd+i
bind[disable_imposter]: cmd+shift+i
bind[connect]: alt
bind[disconnect]: shift+alt
bind[group]: cmd+g
bind[ungroup]: cmd+shift+g
bind[box_select]: b
bind[deselect_all]: q
bind[select_all]: cmd+a
bind[select_similar]: cmd+shift+a
bind[add_to_selection]: shift
bind[bake_gi]: cmd+j
bind[print_objects]: f11
bind[toggle_level_load_stress]: f10

Tools/changelog
- Fixes a bug where openAL break from NaN value in camera.

- Disable the FAQ button in the main menu.

- Adds extra check for null when printing object overwrite warning

- Adds -l flag for quickloading a level.

- Fix so that lipsync searches for files based on relative path.

- Improved debug output for incorrect ref counting.

- Update windows crash message to contain more instructions

- Changes so that instead of deduplication of object in the same position we give the user a warning when saving a level

- Adding callback for JavaScript console messages to the log.

- Moves graphics warning to log instead of popupdialog

- Adds Log() function in AS (with file and line info)

- Makes map_id_ clear on dispose to prevent Get of invalid object pointers.

- Make inherited class constructor virtual to ensure correct execution

- Make unbind explicit to prevent accidental skip for rebind on reused id.
* This fixes a crash issue when opening the menu in the arena level.

- Strips project from alut libraries

- Removes alut dependency by replacing wav loading in filters.

- Fix OpenAL crash on exit (for linux atleast).

- Adds print warning for NONE volume request in skeleton

- JSON support for c++ and AngelScript.

- Converted all previous prints to logging system output

- Adds a logging utility with support for multiplexing and different debug levels.

- Adds mod folder support

- Adds routine to check for particular model loading error presented in FB #4289

- Fixes IEEE NaN generation in terrain normal FB #4303

- Adds a half-fix for bloodsurface texture binding crash by checking id.

- Adds --version flag to game binary

- Makes lack of extra_data_path a non fatal error.

- Adds max test level used for mod integration.

- Makes the FindFilePath more leanient regarding the trailing slash

- Removes version.xml use in crashbugreport for win32
* Replaced with internal version string.

- Moves away from version.xml file in main menu.

- Exposes git version number and build timestamp to AS.
The following functions were added to AS.
GetBuildVersion()
GetBuildTimestamp()

- Adds DebugDrawPoint to AS interface

- Repairs DebugDrawPoint by GL 2.1 -> GL 3.2

- Adds additional control to debug draw text (screenspace and scale)

- Replace old billboard with new generated geom version.

- Adds clip space size locked billboard rendering (for text)

- Adds DebugDrawCircle in AS.

- Fixes DebugDrawWireMesh rendering

- Simple shadow map is updated when objects are added or deleted

- Make sure as_context is valid before updating in Hotspot

- Reduce drawing for debug cylinder, box and sphere to single codepath.

- Adds cylinder debug drawing.

- Adds git version code generation for windows.

- Adds git revision to build (UNIX only atm)

- Adds rain and snow particle systems as hotspots

- Can load dds files if tga is missing

- Terrain uses averagecolor class

- Correctly catch 'missing font' errors

- Improved nav mesh rendering
* The geometry has a brown fill light from the camera instead of being
the normal of the surface.
* The nav mesh surface is stippled in different patterns depending on
distance from camera to give a sense of distance and the ability to more
clearly see through parts of the nav mesh.
* Nav mesh surface has a distance-based linear interpolation between two
colors to make distant parts of the nave mesh less visually prominent.

- Adds finer control of nav mesh in editor settings.

- Adds thick lines for navmesh edges.

- Adds edge rendering and stipple

- Adds basic rendering of nav mesh

- Fixes rendering loaded navmesh, and missing Levels folder issue.
* Levels folder missing was causing navmesh saving to failed.
* Shifts into rendering InputGeom mesh rather than NavMesh mesh to allow
  loaded mesh to render correctly.
* Shifts nav mesh loading into the Draw call as NavMesh loading can be
  threaded sometimes, causing a secondary thread to call opengl.


Source/Internal/config.cpp
//-----------------------------------------------------------------------------
//           Name: config.cpp
//      Developer: Wolfire Games LLC
//    Description:
//        License: Read below
//-----------------------------------------------------------------------------
//
//   Copyright 2022 Wolfire Games LLC
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.
//
//-----------------------------------------------------------------------------
#include "config.h"

#include <Internal/config.h>
#include <Internal/path_utility.h>
#include <Internal/filesystem.h>
#include <Internal/datemodified.h>

#include <Compat/fileio.h>
#include <Logging/logdata.h>
#include <Main/engine.h>

#include <Graphics/textures.h>
#include <Graphics/camera.h>

#include <vector>
#include <algorithm>
#include <iostream>

Config config;
Config default_config;

const vec3 kRedBloodColor(0.4f, 0.0f, 0.0f);
const vec3 kGreenBloodColor(0.0f, 0.4f, 0.0f);
const vec3 kCyanBloodColor(0.0f, 0.4f, 0.4f);
const vec3 kBlackBloodColor(0.1f, 0.1f, 0.1f);

extern bool g_no_reflection_capture;

// Utility function to trim whitespace off the ends of a string
inline std::string trim(std::string source) {
    std::string result = source.erase(source.find_last_not_of(" \t\r") + 1);
    return result.erase(0, result.find_first_not_of(" \t\r"));
}

Config::Config() : has_changed_since_save(false) {
    commonResolutions.push_back(Resolution(1920, 1080));
    commonResolutions.push_back(Resolution(2560, 1440));
    commonResolutions.push_back(Resolution(1366, 768));
    commonResolutions.push_back(Resolution(3840, 2160));
    commonResolutions.push_back(Resolution(1680, 1050));
    commonResolutions.push_back(Resolution(1600, 900));
    commonResolutions.push_back(Resolution(1440, 900));
    commonResolutions.push_back(Resolution(1920, 1200));
    commonResolutions.push_back(Resolution(2560, 1080));
    commonResolutions.push_back(Resolution(1360, 768));
    commonResolutions.push_back(Resolution(1280, 800));
    commonResolutions.push_back(Resolution(1536, 864));
    commonResolutions.push_back(Resolution(2736, 1824));
    commonResolutions.push_back(Resolution(1280, 720));
}

bool Config::Load(const std::string& filename, bool just_filling_blanks, bool shadow_variables) {
    const std::string& configFile = filename;  // pathUtility::localPathToGlobal(filename);
    std::ifstream file;
    my_ifstream_open(file, configFile.c_str(), std::ios_base::in);

    if (!file.is_open())
        return false;

    Load(file, just_filling_blanks, shadow_variables);

    file.close();

    if (just_filling_blanks == false && shadow_variables == false) {
        date_modified_ = GetDateModifiedInt64(configFile.c_str());
        primary_path_ = configFile;
    }

    return true;
}

bool Config::Load(std::istream& stream, bool just_filling_blanks, bool shadow_variables) {
    std::string line;
    std::string comment = "//";
    std::string delimiter = ":";

    Map* _map;

    if (shadow_variables) {
        _map = &shadow_map_;
        LOGI << "Shadow loading" << std::endl;
    } else {
        _map = &map_;
    }

    int count = 0;
    while (stream.good()) {
        getline(stream, line);

        // Remove any comments
        size_t commIdx = line.find(comment);
        if (commIdx != std::string::npos)
            line = line.substr(0, commIdx);

        size_t delimIdx = line.find(delimiter);
        if (delimIdx == std::string::npos)
            continue;

        std::string key = trim(line.substr(0, delimIdx));
        std::string value = trim(line.substr(delimIdx + 1));

        ConfigVal val;
        val.data = value;
        val.order = count;
        ++count;

        if (!key.empty()) {
            // We compare with global map on purpose here, because that is the "true" state.
            if (!just_filling_blanks || map_.find(key) == map_.end()) {
                (*_map)[key] = val;
            }
        }
    }
    return true;
}

static void AddResolution(std::vector<Resolution>& resolutions, int w, int h) {
    // Check if resolution is already there
    int num_res = resolutions.size();
    for (int i = 0; i < num_res; ++i) {
        if (resolutions[i].w == w && resolutions[i].h == h) {
            return;
        }
    }

    resolutions.push_back(Resolution(w, h));
}

Resolution::Resolution(int _w, int _h) : w(_w), h(_h) {
}

int Config::GetMonitorCount() {
    return SDL_GetNumVideoDisplays();
}

std::vector<Resolution> Config::GetPossibleResolutions() {
    int targetMonitor = GetRef("target_monitor").toNumber<int>();
    int displayModeCount = SDL_GetNumDisplayModes(targetMonitor);

    // Populate resolution list
    std::vector<Resolution> resolutions;
    resolutions.reserve(displayModeCount);

    SDL_DisplayMode desktopDisplayMode;
    SDL_GetDesktopDisplayMode(targetMonitor, &desktopDisplayMode);

    float desktopAspect = desktopDisplayMode.w / (float)desktopDisplayMode.h;

    for (int i = 0; i < displayModeCount; ++i) {
        SDL_DisplayMode mode;
        SDL_GetDisplayMode(targetMonitor, i, &mode);

        float resolutionAspect = mode.w / (float)mode.h;

        if ((mode.h <= desktopDisplayMode.h && mode.w <= desktopDisplayMode.w && std::fabs(resolutionAspect - desktopAspect) < 0.01f) || static_cast<FullscreenMode::Mode>(config["fullscreen"].toNumber<int>()) == FullscreenMode::kFullscreen) {
            bool resolutionFound = false;
            for (auto& commonResolution : commonResolutions) {
                if (commonResolution.w == mode.w && commonResolution.h == mode.h) {
                    resolutionFound = true;
                    break;
                }
            }

            if (!resolutionFound)
                continue;

            AddResolution(resolutions, mode.w, mode.h);
        }
    }

    // Add at least this many resolutions to the list in case some user has some weird monitor
    const static int MIN_RESOLUTION_COUNT = 5;
    for (int i = resolutions.size(); resolutions.size() < MIN_RESOLUTION_COUNT && i < displayModeCount; ++i) {
        SDL_DisplayMode mode;
        SDL_GetDisplayMode(targetMonitor, i - resolutions.size(), &mode);

        if (mode.h <= desktopDisplayMode.h && mode.w <= desktopDisplayMode.w) {
            AddResolution(resolutions, mode.w, mode.h);
        }
    }

    // Sort resolutions
    std::sort(resolutions.begin(), resolutions.end(), ResolutionCompare());
    return resolutions;
}

Config* Config::GetPresets() {
    static Config global_settings[4];
    if (global_settings[0].map_.empty()) {
        global_settings[0].GetRef("blood") = 1;
        global_settings[0].GetRef("texture_reduce") = 1;
        global_settings[0].GetRef("multisample") = 1;
        global_settings[0].GetRef("anisotropy") = 2;
        global_settings[0].GetRef("simple_fog") = 1;
        global_settings[0].GetRef("depth_of_field") = 0;
        global_settings[0].GetRef("depth_of_field_reduced") = 1;
        global_settings[0].GetRef("detail_objects") = 0;
        global_settings[0].GetRef("detail_object_decals") = 0;
        global_settings[0].GetRef("detail_object_lowres") = 1;
        global_settings[0].GetRef("detail_object_disable_shadows") = 1;
        global_settings[0].GetRef("detail_objects_reduced") = 1;
        global_settings[0].GetRef("simple_shadows") = 1;
        global_settings[0].GetRef("tet_mesh_lighting") = 0;
        global_settings[0].GetRef("motion_blur_amount") = 0;
        global_settings[0].GetRef("no_reflection_capture") = 1;
        global_settings[0].GetRef("particle_field") = 0;
        global_settings[0].GetRef("particle_field_simple") = 1;
        global_settings[0].GetRef("simple_water") = 1;

        global_settings[1].GetRef("blood") = 2;
        global_settings[1].GetRef("texture_reduce") = 0;
        global_settings[1].GetRef("multisample") = 1;
        global_settings[1].GetRef("anisotropy") = 4;
        global_settings[1].GetRef("simple_fog") = 1;
        global_settings[1].GetRef("depth_of_field") = 1;
        global_settings[1].GetRef("depth_of_field_reduced") = 1;
        global_settings[1].GetRef("detail_objects") = 1;
        global_settings[1].GetRef("detail_object_decals") = 0;
        global_settings[1].GetRef("detail_object_lowres") = 1;
        global_settings[1].GetRef("detail_object_disable_shadows") = 1;
        global_settings[1].GetRef("detail_objects_reduced") = 1;
        global_settings[1].GetRef("simple_shadows") = 1;
        global_settings[1].GetRef("tet_mesh_lighting") = 0;
        global_settings[1].GetRef("motion_blur_amount") = 0;
        global_settings[1].GetRef("no_reflection_capture") = 0;
        global_settings[1].GetRef("particle_field") = 1;
        global_settings[1].GetRef("particle_field_simple") = 1;
        global_settings[1].GetRef("simple_water") = 0;

        global_settings[2].GetRef("blood") = 2;
        global_settings[2].GetRef("texture_reduce") = 0;
        global_settings[2].GetRef("multisample") = 1;
        global_settings[2].GetRef("anisotropy") = 4;
        global_settings[2].GetRef("simple_fog") = 1;
        global_settings[2].GetRef("depth_of_field") = 1;
        global_settings[2].GetRef("depth_of_field_reduced") = 0;
        global_settings[2].GetRef("detail_objects") = 1;
        global_settings[2].GetRef("detail_object_decals") = 0;
        global_settings[2].GetRef("detail_object_lowres") = 1;
        global_settings[2].GetRef("detail_object_disable_shadows") = 1;
        global_settings[2].GetRef("detail_objects_reduced") = 0;
        global_settings[2].GetRef("simple_shadows") = 1;
        global_settings[2].GetRef("tet_mesh_lighting") = 0;
        global_settings[2].GetRef("motion_blur_amount") = 0;
        global_settings[2].GetRef("no_reflection_capture") = 0;
        global_settings[2].GetRef("particle_field") = 1;
        global_settings[2].GetRef("particle_field_simple") = 0;
        global_settings[2].GetRef("simple_water") = 0;

        global_settings[3].GetRef("blood") = 2;
        global_settings[3].GetRef("texture_reduce") = 0;
        global_settings[3].GetRef("multisample") = 4;
        global_settings[3].GetRef("anisotropy") = 4;
        global_settings[3].GetRef("simple_fog") = 0;
        global_settings[3].GetRef("depth_of_field") = 1;
        global_settings[3].GetRef("depth_of_field_reduced") = 0;
        global_settings[3].GetRef("detail_objects") = 1;
        global_settings[3].GetRef("detail_object_lowres") = 0;
        global_settings[3].GetRef("detail_object_decals") = 0;
        global_settings[3].GetRef("detail_object_disable_shadows") = 0;
        global_settings[3].GetRef("detail_objects_reduced") = 0;
        global_settings[3].GetRef("simple_shadows") = 0;
        global_settings[3].GetRef("tet_mesh_lighting") = 0;
        global_settings[3].GetRef("motion_blur_amount") = 0;
        global_settings[3].GetRef("no_reflection_capture") = 0;
        global_settings[3].GetRef("particle_field") = 1;
        global_settings[3].GetRef("particle_field_simple") = 0;
        global_settings[3].GetRef("simple_water") = 0;
    }
    return global_settings;
}

void Config::SetSettingsToPreset(std::string preset_name) {
    int index = 0;
    if (preset_name == "Low") {
        index = 0;
    } else if (preset_name == "Reduced") {
        index = 1;
    } else if (preset_name == "Medium") {
        index = 2;
    } else if (preset_name == "High") {
        index = 3;
    } else {
        return;
    }
    Config::Map& map = GetPresets()[index].map_;
    for (auto& iter : map) {
        config.GetRef(iter.first) = iter.second.data;
    }
}

std::string Config::GetSettingsPreset() {
    int preset = 0;

    // Determine if we match any of the global_settings presets
    for (int i = 0; i < 4; ++i) {
        Config::Map& map = GetPresets()[i].map_;
        for (auto& iter : map) {
            if (config.GetRef(iter.first) != iter.second.data) {
                preset = i + 1;
                break;
            }
        }
        if (preset == i) {
            break;
        }
    }
    if (preset == 4) {
        preset = 0;
    } else {
        ++preset;
    }
    switch (preset) {
        case 0:
            return "Custom";
        case 1:
            return "Low";
        case 2:
            return "Reduced";
        case 3:
            return "Medium";
        case 4:
            return "High";
        default:
            return "Error";
    }
}

std::vector<std::string> Config::GetSettingsPresets() {
    std::vector<std::string> presets;
    // presets.push_back("Custom");
    presets.push_back("Low");
    presets.push_back("Reduced");
    presets.push_back("Medium");
    presets.push_back("High");
    return presets;
}

static const char* difficulty_presets[] = {
    "Casual",
    "Hardcore",
    "Expert",
    NULL};

static const float difficulty_preset_values[] = {
    0.0f, 0.8f,
    0.5f, 1.0f,
    1.0f, 1.0f};

static const bool difficulty_tutorials_preset_values[] = {
    true,
    true,
    false};

static const bool difficulty_ledge_grab_preset_values[] = {
    true,
    true,
    false};

std::vector<std::string> Config::GetDifficultyPresets() {
    std::vector<std::string> presets;
    // presets.push_back("Custom");
    for (int i = 0; difficulty_presets[i] != NULL; i++) {
        presets.push_back(difficulty_presets[i]);
    }
    return presets;
}

std::string Config::GetDifficultyPreset() {
    for (int i = 0; difficulty_presets[i] != NULL; ++i) {
        if (GetRef("game_difficulty").toNumber<float>() == difficulty_preset_values[i * 2] &&
            GetRef("global_time_scale_mult").toNumber<float>() == difficulty_preset_values[i * 2 + 1] &&
            GetRef("tutorial").toBool() == difficulty_tutorials_preset_values[i] &&
            GetRef("auto_ledge_grab").toBool() == difficulty_ledge_grab_preset_values[i]) {
            return difficulty_presets[i];
        }
    }
    return "Custom";
}

// Used to validate progress, rounds down in difficulty steps
std::string Config::GetClosestDifficulty() {
    std::string difficulty = "Custom";
    for (int i = 0; difficulty_presets[i] != NULL; ++i) {
        if (GetRef("game_difficulty").toNumber<float>() >= difficulty_preset_values[i * 2] &&
            GetRef("global_time_scale_mult").toNumber<float>() >= difficulty_preset_values[i * 2 + 1]) {
            difficulty = difficulty_presets[i];
        } else {
            break;
        }
    }
    return difficulty;
}

void Config::SetDifficultyPreset(std::string name) {
    for (int i = 0; difficulty_presets[i] != NULL; i++) {
        if (strmtch(difficulty_presets[i], name.c_str())) {
            GetRef("game_difficulty") = difficulty_preset_values[i * 2];
            GetRef("global_time_scale_mult") = difficulty_preset_values[i * 2 + 1];
            GetRef("tutorials") = difficulty_tutorials_preset_values[i];
            GetRef("auto_ledge_grab") = difficulty_ledge_grab_preset_values[i];
        }
    }
}

void Config::ReloadDynamicSettings() {
    Engine::Instance()->SetGameSpeed(config["global_time_scale_mult"].toNumber<float>(), true);
    Engine::Instance()->GetSound()->SetMusicVolume(config.GetRef("music_volume").toNumber<float>());
    Engine::Instance()->GetSound()->SetMasterVolume(config.GetRef("master_volume").toNumber<float>());
    Graphics::Instance()->config_.motion_blur_amount_ = (float)config["motion_blur_amount"].toNumber<int>();
    Input::Instance()->SetMouseSensitivity(config["mouse_sensitivity"].toNumber<float>());
    Input::Instance()->UpdateGamepadLookSensitivity();
    Input::Instance()->UpdateGamepadDeadzone();
}

void Config::ReloadStaticSettings() {
    Input::Instance()->SetFromConfig(config);
    Graphics::Instance()->SetFromConfig(config, true);
    Graphics::Instance()->SetTargetMonitor(config["target_monitor"].toNumber<int>());
    Graphics::Instance()->SetResolution(config["screenwidth"].toNumber<int>(), config["screenheight"].toNumber<int>(), false);
    Graphics::Instance()->SetFSAA(config["multisample"].toNumber<int>());
    if (config["anisotropy"].toNumber<int>() != Graphics::Instance()->config_.anisotropy()) {
        Graphics::Instance()->SetAnisotropy((float)config["anisotropy"].toNumber<int>());
        Textures::Instance()->ApplyAnisotropy();
    }
    Graphics::Instance()->SetSimpleFog(config["simple_fog"].toBool());
    Graphics::Instance()->SetDepthOfField(config["depth_of_field"].toBool());
    Graphics::Instance()->SetDepthOfFieldReduced(config["depth_of_field_reduced"].toBool());
    Graphics::Instance()->SetDetailObjects(config["detail_objects"].toBool());
    Graphics::Instance()->SetDetailObjectDecals(config["detail_object_decals"].toBool());
    Graphics::Instance()->SetDetailObjectLowres(config["detail_object_lowres"].toBool());
    Graphics::Instance()->SetDetailObjectShadows(!config["detail_object_disable_shadows"].toBool());
    Graphics::Instance()->SetDetailObjectsReduced(config["detail_objects_reduced"].toBool());
    Graphics::Instance()->setAttribEnvObjInstancing(config["attrib_envobj_instancing"].toBool());
    // Makes the game crash
    // the_scenegraph->light_probe_collection.light_volume_enabled = config["light_volume_lighting"].toBool();
    Graphics::Instance()->SetFullscreen(static_cast<FullscreenMode::Mode>(config["fullscreen"].toNumber<int>()));
    Graphics::Instance()->SetVsync(config["vsync"].toBool());
    Graphics::Instance()->config_.SetBlood(config.GetRef("blood").toNumber<int>());
    Graphics::Instance()->config_.SetBloodColor(GraphicsConfig::BloodColorFromString(config["blood_color"].str()));
    Graphics::Instance()->config_.SetSplitScreen(config["split_screen"].toNumber<bool>());
    Graphics::Instance()->SetSeamlessCubemaps(config["seamless_cubemaps"].toNumber<bool>());
    g_no_reflection_capture = config["no_reflection_capture"].toNumber<bool>();

    Input::Instance()->SetInvertXMouseLook(config["invert_x_mouse_look"].toNumber<bool>());
    Input::Instance()->SetInvertYMouseLook(config["invert_y_mouse_look"].toNumber<bool>());
    Input::Instance()->UseRawInput(config["use_raw_input"].toNumber<bool>());
    int num_cameras = ActiveCameras::NumCameras();
    int curr_id = ActiveCameras::GetID();
    for (int i = 0; i < num_cameras; ++i) {
        ActiveCameras::Set(i);
        ActiveCameras::Get()->SetAutoCamera(config["auto_camera"].toNumber<bool>());
    }
    ActiveCameras::Set(curr_id);

    SceneGraph* scene_graph = Engine::Instance()->GetSceneGraph();
    if (scene_graph)
        scene_graph->PreloadShaders();
}

class ConfigValCompare {
   public:
    bool operator()(const std::pair<std::string, ConfigVal>& a,
                    const std::pair<std::string, ConfigVal>& b) {
        return a.second.order < b.second.order;
    }
};

bool Config::Save(const std::string& filename) {
    LOGI << "Saving config to: " << filename << std::endl;
    CreateParentDirs(filename.c_str());
    std::ofstream file;
    my_ofstream_open(file, filename);

    if (!file.is_open())
        return false;

    std::vector<std::pair<std::string, ConfigVal> > vec(map_.begin(), map_.end());
    std::sort(vec.begin(), vec.end(), ConfigValCompare());

    for (auto& i : vec) {
        file << i.first << ": " << i.second.data.str() << "\n";
    }

    file.close();

    has_changed_since_save = false;

    if (AreSame(filename.c_str(), primary_path_.c_str())) {
        date_modified_ = GetDateModifiedInt64(primary_path_.c_str());
    }

    return true;
}

const StringVariant& Config::operator[](const std::string& keyName) const {
    std::map<std::string, ConfigVal>::const_iterator iter = shadow_map_.find(keyName);

    if (iter != shadow_map_.end()) {
        return iter->second.data;
    }

    iter = map_.find(keyName);
    if (iter != map_.end()) {
        return iter->second.data;
    }

    static StringVariant empty;
    return empty;
}

bool Config::HasKey(const char* key) {
    std::string s = std::string(key);
    return HasKey(s);
}

bool Config::HasKey(std::string& key) {
    std::map<std::string, ConfigVal>::const_iterator iter = shadow_map_.find(key);

    if (iter != shadow_map_.end()) {
        return true;
    }

    iter = map_.find(key);

    if (iter != map_.end()) {
        return true;
    }

    return false;
}

void Config::RemoveConfig(std::string key) {
    std::map<std::string, ConfigVal>::iterator iter = shadow_map_.find(key);

    if (iter != shadow_map_.end()) {
        shadow_map_.erase(iter);
    }

    iter = map_.find(key);

    if (iter != map_.end()) {
        map_.erase(iter);
    }
}

bool Config::HasChangedSinceLastSave() {
    return has_changed_since_save;
}

bool Config::operator!=(const Config& other) const {
    return !(*this == other);
}

bool Config::operator==(const Config& other) const {
    return map_ == other.map_;
}

bool Config::PrimarySourceModified() {
    if (primary_path_.empty() == false) {
        return GetDateModifiedInt64(primary_path_.c_str()) != date_modified_;
    } else {
        return false;
    }
}

std::vector<Resolution>& Config::GetCommonResolutions() {
    return commonResolutions;
}

bool ConfigVal::operator==(const ConfigVal& other) const {
    return data == other.data;
}

bool StringVariant::operator==(const StringVariant& other) const {
    return data == other.data;
}

bool StringVariant::operator!=(const StringVariant& other) const {
    return !((*this) == other);
}
